File: providers/openrouter/embed.go
package openrouter

import (
	"context"
	"fmt"

	"github.com/mltheuser/ai-router/api"
)

// embedRequest is the OpenRouter-specific embedding request (OpenAI-compatible).
type embedRequest struct {
	Model      string   `json:"model"`
	Input      []string `json:"input"`
	Dimensions     *int     `json:"dimensions,omitempty"`
}

// embedResponse is the OpenRouter-specific embedding response (OpenAI-compatible).
type embedResponse struct {
	Object string          `json:"object"`
	Data   []embedData     `json:"data"`
	Model  string          `json:"model"`
	Usage  embedUsage      `json:"usage"`
}

type embedData struct {
	Object    string    `json:"object"`
	Embedding []float64 `json:"embedding"`
	Index     int       `json:"index"`
}

type embedUsage struct {
	PromptTokens int `json:"prompt_tokens"`
	TotalTokens  int `json:"total_tokens"`
}

// Embed generates embeddings via OpenRouter (OpenAI-compatible endpoint).
func (p *Provider) Embed(ctx context.Context, req *api.EmbedRequest) (*api.EmbedResponse, error) {
	// Build provider-specific request
	// Build provider-specific request
	provReq := embedRequest{
		Model:      req.Model,
		Input:      req.Input,
		Dimensions: req.Dimensions,
	}

	var provResp embedResponse
	if err := p.client.post(ctx, "/embeddings", provReq, &provResp); err != nil {
		return nil, fmt.Errorf("openrouter embed: %w", err)
	}

	// Convert to unified response
	resp := &api.EmbedResponse{
		Object: provResp.Object,
		Model:  provResp.Model,
		Usage: api.EmbedUsage{
			PromptTokens: provResp.Usage.PromptTokens,
			TotalTokens:  provResp.Usage.TotalTokens,
		},
	}

	for _, d := range provResp.Data {
		resp.Data = append(resp.Data, api.EmbedData{
			Object:    d.Object,
			Embedding: d.Embedding,
			Index:     d.Index,
		})
	}

	return resp, nil
}

File: SDKs/kotlin/src/main/kotlin/ai/router/sdk/models/Embed.kt
package ai.router.sdk.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * OpenAI-compatible embedding request.
 */
@Serializable
public data class EmbedRequest(
    val model: String,
    val input: List<String>,
    val dimensions: Int? = null,
)

/**
 * OpenAI-compatible embedding response.
 */
@Serializable
public data class EmbedResponse(
    val `object`: String,
    val data: List<EmbedData>,
    val model: String,
    val usage: EmbedUsage,
)

@Serializable
public data class EmbedData(
    val `object`: String,
    val embedding: List<Double>,
    val index: Int,
)

@Serializable
public data class EmbedUsage(
    @SerialName("prompt_tokens") val promptTokens: Int,
    @SerialName("total_tokens") val totalTokens: Int,
)
